src/NuGet.Core/NuGet.DependencyResolver.Core/GraphModel/GraphOperations.cs

```diff
@@ -541,15 +548,18 @@ public static void ForEach<TItem>(this GraphNode<TItem> root, Action<GraphNode<T
             while (queue.Count > 0)
             {
                 var node = queue.Dequeue();
-                visitor(node);
+                if (skipNode == null || !skipNode(node))
+                {
+                    visitor(node);
 
-                AddInnerNodesToQueue(node.InnerNodes, queue);
+                    AddInnerNodesToQueue(node.InnerNodes, queue);
+                }
             }
 
             Cache<TItem>.ReleaseQueue(queue);
         }
 
-        public static void ForEach<TItem, TContext>(this GraphNode<TItem> root, Action<GraphNode<TItem>, TContext> visitor, TContext context)
+        public static void ForEach<TItem, TContext>(this GraphNode<TItem> root, Action<GraphNode<TItem>, TContext> visitor, TContext context, Func<GraphNode<TItem>, bool> skipNode = null)
```

Comment:
This is a public API change, please create an overload.

Reply:
Updated.